rubix-common/src/main/java/com/qubole/rubix/common/metrics/BookKeeperMetricsFilter.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Copyright (c) 2018. Qubole Inc
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License. See accompanying LICENSE file.
+ */
+package com.qubole.rubix.common.metrics;
+
+import com.codahale.metrics.Metric;
+import com.codahale.metrics.MetricFilter;
+import com.qubole.rubix.spi.CacheConfig;
+import org.apache.hadoop.conf.Configuration;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class BookKeeperMetricsFilter implements MetricFilter
+{
+  private final Configuration conf;
+
+  public BookKeeperMetricsFilter(Configuration conf)
+  {
+    this.conf = conf;
+  }
+
+  @Override
+  public boolean matches(String name, Metric metric)
+  {
+    List<String> whitelist = new ArrayList<>();
+    if (CacheConfig.areLivenessMetricsEnabled(conf)) {
+      whitelist.addAll(BookKeeperMetrics.LivenessMetric.getAllNames());
+    }
+    if (CacheConfig.areCacheMetricsEnabled(conf)) {
+      whitelist.addAll(BookKeeperMetrics.CacheMetric.getAllNames());
+    }
+    if (CacheConfig.areJvmMetricsEnabled(conf)) {
+      whitelist.addAll(BookKeeperMetrics.BookKeeperJvmMetric.getAllNames());
+      whitelist.addAll(BookKeeperMetrics.LDTSJvmMetric.getAllNames());
+    }
+
+    for (String whitelistedMetric : whitelist) {
+      if (name.startsWith(whitelistedMetric)) {
+        return true;
+      }
+    }
+    return false;
+  }
+}
```

Comment:
We are getting this exception in the test
```
2018-08-02 00:16:13 [DEBUG] metrics.JmxReporter: Unable to unregister gauge
javax.management.InstanceNotFoundException: metrics:name=rubix.bookkeeper.jvm.threads.new.count
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.getMBean(DefaultMBeanServerInterceptor.java:1095)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.exclusiveUnregisterMBean(DefaultMBeanServerInterceptor.java:427)
	at com.sun.jmx.interceptor.DefaultMBeanServerInterceptor.unregisterMBean(DefaultMBeanServerInterceptor.java:415)
	at com.sun.jmx.mbeanserver.JmxMBeanServer.unregisterMBean(JmxMBeanServer.java:546)
	at com.codahale.metrics.JmxReporter$JmxListener.unregisterMBean(JmxReporter.java:521)
	at com.codahale.metrics.JmxReporter$JmxListener.onGaugeRemoved(JmxReporter.java:543)
	at com.codahale.metrics.MetricRegistry.notifyListenerOfRemovedMetric(MetricRegistry.java:366)
	at com.codahale.metrics.MetricRegistry.onMetricRemoved(MetricRegistry.java:360)
	at com.codahale.metrics.MetricRegistry.remove(MetricRegistry.java:156)
	at com.codahale.metrics.MetricRegistry.removeMatching(MetricRegistry.java:170)
	at com.qubole.rubix.bookkeeper.BookKeeperServer.removeMetrics(BookKeeperServer.java:149)

```

The main reason is we are registering all the metrics (JVM, LIVELINESS) in the registerMetrics while starting the server.
But based on the flag, we are only pushing some metrics in JMX through BookKeeperMetricsFilter. So during unregister, the system is trying to remove some metrics (already registered) and failing because those are not available in JMX. (By default the JMX metrics are not pushed).
We need to handle this scenario.